source/Nevermore/Advanced/ReadTransaction.cs

```diff
@@ -637,9 +637,45 @@ void ITransactionDiagnostic.WriteCurrentTransactions(StringBuilder output)
 
         public void Dispose()
         {
-            Transaction?.Dispose();
-            connection?.Dispose();
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        public async ValueTask DisposeAsync()
+        {
+            await DisposeAsyncCore();
+
+            Dispose(false);
+
+#pragma warning disable CA1816 // Dispose methods should call SuppressFinalize
+            GC.SuppressFinalize(this);
+#pragma warning restore CA1816 // Dispose methods should call SuppressFinalize
+        }
+
+        protected virtual void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                Transaction?.Dispose();
+                connection?.Dispose();
+            }
+
             registry.Remove(this);
+
+            Transaction = null;
+            connection = null;
+        }
+
+        protected virtual async ValueTask DisposeAsyncCore()
+        {
+            if(Transaction != null)
+                await Transaction.DisposeAsync().ConfigureAwait(false);
+
+            if(connection != null)
+                await connection.DisposeAsync().ConfigureAwait(false);
+
+            Transaction = null;
+            connection = null;
```

Comment:
This pattern for `IDisposable` and `IAsyncDisposable` is from Microsoft Docs: https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-disposeasync#implement-both-dispose-and-async-dispose-patterns